### PR TITLE
Semigroup and monoid

### PR DESCRIPTION
**Pull request by harmenjanssen**
Semigroup and Monoid are coming.
Semigroup is in here, but Monoid is so closely related it will follow soon. 

The concrete implementations will derive from `Monoid` as well, in the end.   
I'm not sure about the `Impl` namespace, but have yet to come up with a better namespace (the problem being there isn't really an appropriate namespace grouping everyone that will be in there). 

Also, I'm assuming a lot more documentation is necessary (maybe in the form of a blog post) to explain the concrete Monoid-implementations.

### CODE REVIEW

File: types/Impl/Product.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+use Garp\Functional\Types\{Semigroup, Setoid};
+
+/**
+ * @package  Garp\Functional
+ * @author   Harmen Janssen <harmen@grrr.nl>
+ * @license  https://github.com/grrr-amsterdam/garp-functional/blob/master/LICENSE.md BSD-3-Clause
+ * @see      https://github.com/fantasyland/fantasy-land#semigroup
+ */
+class Product implements Semigroup, Setoid {
```
Comment (martijngastkemper) [inline types/Impl/Product.php]: Might be good to make `Type\Impl` classes `final` to force using the type interfaces.

File: types/Impl/Product.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?php
+declare(strict_types=1);
+
+namespace Garp\Functional\Types\Impl;
```
Comment (martijngastkemper) [inline types/Impl/Product.php]: Do these classes actually need a separate namespace? And if they do, maybe create a new namespace: `Garp\Functional\Impl` (discussion about the name stil continues) besides `Garp\Functional\Types`.

**Review by martijngastkemper — COMMENTED** [review]
Nice addition!

File: types/Impl/Product.php (hunk 2 above)
Comment (harmenjanssen) [reply types/Impl/Product.php]: Well, I would just like to group them _somewhere_... they are distinctly different from everything else in this library. 
But it's a good idea to pull them out of `Types`. 👍

File: types/Impl/Product.php (hunk 1 above)
Comment (harmenjanssen) [reply types/Impl/Product.php]: Yes, good idea, will do. 👍

**Comment by harmenjanssen** [comment]
I'm really curious what you think of the `fold` function. It's purpose is to work with Monoids. 
A Monoid is a typeclass like Semigroup, but with the idea of an identity value.

For example, addition is a Monoid, because:

```
5 + 10 = 15 // concatenation
5 + 0 = 5 // identity value = 0
0 + 5 = 5 // left identity and right identity are the most important rules of a Monoid, 
          // it doesn't matter on which end the identity value is concatenated
```

So it's a lot like reduce.


----

My implementation works, but I don't like the first argument being a string containing the classname. 
It's clumsy to typecheck, but in PHP I cannot pass around a reference to a class.

---

In Haskell it's discovered from the type signature. `mempty` is the function that I've defined as `Monoid::empty`. 

Given these two functions, which have the same implementation (namely they are simple aliases for `mempty`):

```haskell
foo :: (Num a) => Int -> (Sum a)
foo = mempty

bar :: (Num a) => Int -> (Product a)
bar = mempty
```

it will simply determine from the signature it needs to return a `Sum` when calling `foo 42`, and a `Product` when calling `bar 42`.
(the signature is the line above the function, containing `::`)


Anyway, that's just some context, it's not possible in PHP, so no reason to dwell on it. 🙂 

An interesting implementation is this one though, in PHP: https://github.com/i-am-tom/php-folding-talk/blob/master/src/Fold.php in which a Monoid factory object is required as parameter. 
It makes it easier to type check, but for a library like Garp\Functional pragmatism should be a huge factor – we want it to improve your productivity. I'm not sure whether requiring a second class simply to provide a factory for the first class is the way to go for us.

**Comment by harmenjanssen** [comment]
Maybe we could play around with reflection by the way, but I'm never sure of its impact on performance to check function signatures/docblocks at runtime.

**Comment by martijngastkemper** [comment]
You already wrote down my opinion:

> it's not possible in PHP, so no reason to dwell on it. 🙂

The addition of `::class` made it possible to programmatically get the class string. Replacing it by reflection is not worth the performance costs in my opinion (based on some quick reading on Stack Overflow). The other implementation you send doesn't feel like Garp Functional.

Beside a blog post about Semigroup/Monoids. It's possible to write one in which we explain why we choose string instead of reflection. I assume that a benchmark will make clear that the string approach is more performant.

**Comment by harmenjanssen** [comment]
👍 Thanks for the feedback! 
Will extend this with other implementations shortly, after which this can be merged.

**Comment by harmenjanssen** [comment]
I think I'm finally happy with this implementation.

I ended up moving the typeclasses to `Garp\Functional\Types\TypeClasses`: it's just more in line with existing lingo (see also: https://www.quora.com/What-is-the-difference-between-a-type-and-a-type-class-in-Haskell). 
It requires a major version bump because `Setoid` and `Ord` will no longer be in the same place, but I think it's worth it, since we're still in the early stages.

Let me know what you think. 🙂

File: tests/types/ProductTest.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            function ($out, $n) {
+                $a = rand(0, 1000);
+                $b = rand(0, 1000);
+                $product = $a * $b;
+                $out[] = [$a, $b, $product];
+                return $out;
+            },
+            [],
+            range(0, 100)
+        );
```
Comment (martijngastkemper) [inline tests/types/ProductTest.php]: Isn't it an option to use `repeat` instead of `reduce`? Doesn't really matter, but just curiosity.

**Review by martijngastkemper — APPROVED** [review]
LGTM! I first didn't get the TypeClass, because they are all interfaces. But that is wat het exactly is in a PHP context.

Don't forget to remove `substract` in this major version.

File: tests/types/ProductTest.php (hunk 3 above)
Comment (harmenjanssen) [reply tests/types/ProductTest.php]: Definitely, that makes the intent more clear, thanks 👍

**Pull request merged**